src/main/kotlin/app/CacheService.kt

```diff
@@ -0,0 +1,98 @@
+@file:Suppress("SqlResolve")
+
+package app
+
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.databind.DeserializationContext
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer
+import com.fasterxml.jackson.databind.module.SimpleModule
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+import org.slf4j.LoggerFactory
+import java.sql.DriverManager
+import java.time.LocalDateTime
+import java.time.temporal.ChronoUnit
+import java.util.Date
+
+object CacheService {
+    private const val urlToDb = "jdbc:h2:mem:userinfo"
+    private val log = LoggerFactory.getLogger(CacheService.javaClass)
+
+    private fun createTableIfAbsent() {
+        val connection = DriverManager.getConnection(urlToDb)
+        val statement = connection.createStatement()
+
+        statement.execute(
+            "CREATE TABLE IF NOT EXISTS userinfo (" +
+            "id VARCHAR2 PRIMARY KEY," +
+            "timestamp TIMESTAMP WITH TIME ZONE, " +
+            "data JSON" +
+            ")"
+        )
+    }
+
+    fun lookUpInCache(username: String): UserProfile? {
+        val connection = DriverManager.getConnection(urlToDb)
+
+        createTableIfAbsent()
+
+        val preparedStatement = connection.prepareStatement(
+            "SELECT " +
+            "timestamp, " +
+            "data " +
+            "FROM userinfo " +
+            "WHERE id = ?"
+        )
+        preparedStatement.setString(1, username)
+
+        val result = preparedStatement.executeQuery()
+        result.use {
+            // guaranteed to be at most one.
+            if (it.next()) {
+                val timestamp = it.getTimestamp(1).toLocalDateTime()
+                val diffInHours = ChronoUnit.HOURS.between(timestamp, LocalDateTime.now())
+                if (diffInHours <= 6) {
+                    val json: String = it.getString(2)
+
+                    log.debug("cache hit: {}", json)
+
+                    val simpleModule = SimpleModule()
+                    simpleModule.addDeserializer(Date::class.java, DateDeserializer())
+
+                    val objectMapper = jacksonObjectMapper()
+                    objectMapper.registerModule(simpleModule)
```

Comment:
I'd just save one instance instead of repeatedly creating the mapper

Reply:
Shame on me!
https://github.com/Attacktive/profile-summary-for-github/commit/5a929a9945073fb3ea937d6c7d93c30a0212c43f